napari/plugins/hook_specifications.py

```diff
@@ -0,0 +1,56 @@
+"""
+All hook specifications for pluggable functionality should be defined here.
+
+A hook specification is a function signature (with documentation) that declares
+an API that plugin developers must adhere to when providing hook
+implementations.  Hook implementations provided by plugins (and internally by
+napari) will then be invoked in various places throughout the code base.
+
+NOTE: in pluggy documentation, hook specification marker instances are named
+"hookspec" by convention, and hook implementation marker instances are named
+"hookimpl".  The convention in napari is to name them more explicity:
+"napari_hook_spcification" and "napari_hook_implementation", respectively.
+
+hook specifications are a feature of pluggy:
+https://pluggy.readthedocs.io/en/latest/#specs
+
+These hook specifications also serve as the API reference for plugin
+developers, so comprehensive documentation with complete type annotations is
+imperative.
+"""
+
+import pluggy
+from typing import Callable, Optional, List, Tuple, Union, Any, Dict
+
+napari_hook_specification = pluggy.HookspecMarker("napari")
+
+# layer data may be: (data,) (data, meta), or (data, meta, layer_type)
+# using "Any" for the data type for now
+LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
+ReaderFunction = Callable[[str], List[LayerData]]
+
+
+@napari_hook_specification(firstresult=True)
+def napari_get_reader(path: str) -> Optional[ReaderFunction]:
+    """Return function capable of loading `path` into napari, or None.
+
+    This function will be called on File -> Open... or when a user drags and
+    drops a file/folder onto the viewer. This function must execute QUICKLY,
```

Comment:
Can you define "QUICKLY" here? ie what's the current timeout?

Reply:
i'd like it to be under ~10-50ms...  but that's up for discussion.
there's no timeout implemented anywhere here... but a strict timeout check is one of the things is offered in #935